File: core/src/main/java/com/ustadmobile/core/contentformats/xapi/endpoints/StateEndpoint.java
package com.ustadmobile.core.contentformats.xapi.endpoints;

import com.google.gson.Gson;
import com.ustadmobile.core.contentformats.xapi.Actor;
import com.ustadmobile.core.contentformats.xapi.State;
import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.db.dao.AgentDao;
import com.ustadmobile.core.db.dao.PersonDao;
import com.ustadmobile.core.db.dao.StateContentDao;
import com.ustadmobile.core.db.dao.StateDao;
import com.ustadmobile.lib.db.entities.AgentEntity;
import com.ustadmobile.lib.db.entities.StateContentEntity;
import com.ustadmobile.lib.db.entities.StateEntity;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import static com.ustadmobile.core.contentformats.xapi.endpoints.XapiUtil.getAgent;
import static com.ustadmobile.core.contentformats.xapi.endpoints.XapiUtil.insertOrUpdateState;
import static com.ustadmobile.core.contentformats.xapi.endpoints.XapiUtil.insertOrUpdateStateContent;

public class StateEndpoint {

    private final UmAppDatabase db;
    private final Gson gson;
    private final AgentDao agentDao;
    private final StateDao stateDao;
    private final StateContentDao stateContentDao;
    private final PersonDao personDao;

    public StateEndpoint(UmAppDatabase db, Gson gson) {
        this.db = db;
        this.gson = gson;

        personDao = db.getPersonDao();
        agentDao = db.getAgentDao();
        stateDao = db.getStateDao();
        stateContentDao = db.getStateContentDao();
    }


    public void storeState(State state) throws IllegalArgumentException {

        StatementEndpoint.checkValidActor(state.getAgent());

        AgentEntity agentEntity = getAgent(agentDao, personDao, state.getAgent());

        StateEntity stateEntity = insertOrUpdateState(stateDao, state, agentEntity.getAgentUid());

        insertOrUpdateStateContent(stateContentDao, state.getContent(), stateEntity);


    }


    public String getStateContent(String stateId) {

        StateEntity entity = stateDao.findByStateId(stateId);
        List<StateContentEntity> list = stateContentDao.findAllStateContentWithStateUid(entity.getStateUid());
        HashMap<String, String> contentMap = new HashMap<>();
        if (list != null) {
            for (StateContentEntity contentEntity : list) {
                contentMap.put(contentEntity.getStateContentKey(), contentEntity.getStateContentValue());
            }
        }

        return gson.toJson(contentMap);
    }

    public void deleteStateContent(String stateId) {

        stateDao.setStateInActive(stateId, false);
    }

    public String getListOfStateId(String agentJson, String activityId, String registration, String since) {

        Actor agent = gson.fromJson(agentJson, Actor.class);

        AgentEntity agentEntity = getAgent(agentDao, personDao, agent);

        List<StateEntity> list = stateDao.findStateIdByAgentAndActivity(agentEntity.getAgentUid(), activityId, registration, since);

        List<String> idList = new ArrayList<>();
        for (StateEntity state : list) {
            idList.add(state.getStateId());
        }

        return gson.toJson(idList);

    }

    public void deleteListOfStates(String agentJson, String activityId, String registration, String since) {

        Actor agent = gson.fromJson(agentJson, Actor.class);

        AgentEntity agentEntity = getAgent(agentDao, personDao, agent);

        stateDao.updateStateToInActive(agentEntity.getAgentUid(), activityId, registration, since, false);
    }
}

File: core/src/main/java/com/ustadmobile/core/contentformats/xapi/endpoints/StateEndpoint.kt
package com.ustadmobile.core.contentformats.xapi.endpoints

import com.google.gson.Gson
import com.ustadmobile.core.contentformats.xapi.Actor
import com.ustadmobile.core.contentformats.xapi.State
import com.ustadmobile.core.contentformats.xapi.endpoints.XapiUtil.getAgent
import com.ustadmobile.core.contentformats.xapi.endpoints.XapiUtil.insertOrUpdateState
import com.ustadmobile.core.contentformats.xapi.endpoints.XapiUtil.insertOrUpdateStateContent
import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.db.dao.AgentDao
import com.ustadmobile.core.db.dao.PersonDao
import com.ustadmobile.core.db.dao.StateContentDao
import com.ustadmobile.core.db.dao.StateDao
import java.util.*

class StateEndpoint(db: UmAppDatabase, private val gson: Gson) {
    private val agentDao: AgentDao = db.agentDao
    private val stateDao: StateDao = db.stateDao
    private val stateContentDao: StateContentDao = db.stateContentDao
    private val personDao: PersonDao = db.personDao


    @Throws(IllegalArgumentException::class)
    fun storeState(state: State) {

        StatementEndpoint.checkValidActor(state.agent!!)

        val agentEntity = getAgent(agentDao, personDao, state.agent!!)

        val stateEntity = insertOrUpdateState(stateDao, state, agentEntity.agentUid)

        insertOrUpdateStateContent(stateContentDao, state.content!!, stateEntity)

    }


    fun getStateContent(stateId: String): String {

        val entity = stateDao.findByStateId(stateId)
        val list = stateContentDao.findAllStateContentWithStateUid(entity.stateUid)
        val contentMap = HashMap<String, String>()
        if (list != null) {
            for (contentEntity in list) {
                contentMap[contentEntity.stateContentKey] = contentEntity.stateContentValue
            }
        }

        return gson.toJson(contentMap)
    }

    fun deleteStateContent(stateId: String) {

        stateDao.setStateInActive(stateId, false)
    }

    fun getListOfStateId(agentJson: String, activityId: String, registration: String, since: String): String {

        val agent = gson.fromJson(agentJson, Actor::class.java)

        val agentEntity = getAgent(agentDao, personDao, agent)

        val list = stateDao.findStateIdByAgentAndActivity(agentEntity.agentUid, activityId, registration, since)

        val idList = ArrayList<String>()
        for (state in list) {
            idList.add(state.stateId)
        }

        return gson.toJson(idList)

    }

    fun deleteListOfStates(agentJson: String, activityId: String, registration: String, since: String) {

        val agent = gson.fromJson(agentJson, Actor::class.java)

        val agentEntity = getAgent(agentDao, personDao, agent)

        stateDao.updateStateToInActive(agentEntity.agentUid, activityId, registration, since, false)
    }
}
